onnxruntime/core/graph/function_utils.cc

```diff
@@ -0,0 +1,672 @@
+  // Get a temporary tensor-shape map
+  const auto num_func_inputs = func_proto.input_size();
+  std::unordered_map<std::string, TypeProto*> value_types_by_name;
+  std::vector<TypeProto> types_cache(func_proto.input_size());
+  for (int i = 0; i < num_func_inputs; ++i) {
+    types_cache[i] = *ctx.getInputType(i);
+    value_types_by_name[func_proto.input().Get(i)] = &types_cache[i];
+  }
+
+  // Get a temporary initial value map
+  std::unordered_map<std::string, const TensorProto*> initializers_by_name;
+  std::unordered_map<std::string, const SparseTensorProto*> sparse_initializers_by_name;
+  for (int i = 0; i < static_cast<int>(ctx.getNumInputs()) && i < num_func_inputs; ++i) {
+    const TypeProto* type = ctx.getInputType(i);
+    if (type->value_case() == TypeProto::kTensorType && ctx.getInputData(i) != nullptr) {
+      initializers_by_name[func_proto.input().Get(i)] = ctx.getInputData(i);
+    } else if (type->value_case() == TypeProto::kSparseTensorType &&
+               ctx.getInputSparseData(i) != nullptr) {
+      sparse_initializers_by_name[func_proto.input().Get(i)] = ctx.getInputSparseData(i);
+    }
+  }
+  std::unordered_map<std::string, const AttributeProto*> attr_map;
+  for (auto& attr : func_proto.attribute()) {
+    if (ctx.getAttribute(attr) != nullptr) {
+      attr_map[attr] = ctx.getAttribute(attr);
+    }
+  }
+
+  for (auto& n : func_proto.node()) {
+    NodeProto copy_n(n);
+    // Add attribute information into the temporary node
+    copy_n.clear_attribute();
+    for (const auto& attr : n.attribute()) {
+      if (attr.has_ref_attr_name()) {
+        if (attr_map.count(attr.ref_attr_name())) {
+          auto copy_attr = *attr_map[attr.ref_attr_name()];
+          copy_attr.set_name(attr.name());
+          copy_n.add_attribute()->CopyFrom(copy_attr);
+        }
+      } else {
+        copy_n.add_attribute()->CopyFrom(attr);
+      }
+    }
+    ONNX_NAMESPACE::shape_inference::InferenceContextImpl func_node_ctx(
+        copy_n, value_types_by_name, initializers_by_name, sparse_initializers_by_name, {});
+
+    // Resolve domain for node
+    auto it = func_opset_imports.find(n.domain());
+    if (it == func_opset_imports.end()) {
+      fail_type_inference("Cannot infer type and shape for function", func_proto.name(),
+                          ". No opset import for domain", n.domain(), " referenced by function body node ",
+                          n.name(), " optype ", n.op_type());
+    }
+    auto domain_version = it->second;
+    const auto schema = schema_registry->GetSchema(n.op_type(), domain_version, n.domain());
+    if (schema) {
+      schema->GetTypeAndShapeInferenceFunction()(func_node_ctx);
+    } else {
+      // check model local functions for FunctionProto
+      auto iter = in_model_functions.find(get_func_id(n.domain(), n.op_type()));
+      if (iter == in_model_functions.end()) {
+        return;
+      }
+
+      std::unordered_map<std::string, int> func_node_opset_imports;
```

Comment:
> std::unordered_map<std::string, int> func_node_opset_imports; [](http://example.com/codeflow?start=5&length=62)

Lots of maps. Each std node container allocates a node on default construction even it is not used.